### PR TITLE
Improve support of custom data

### PR DESCRIPTION
**Pull request by kalletlak**
Frontend changes related to
https://github.com/cBioPortal/cbioportal/issues/7726
https://github.com/cBioPortal/cbioportal/issues/6482
https://github.com/cBioPortal/cbioportal/issues/7609
https://github.com/cBioPortal/cbioportal/issues/8044
https://github.com/cBioPortal/cbioportal/issues/8057
https://github.com/cBioPortal/cbioportal/issues/8218

To test
https://cbioportal-fix-7726-bs0ipkcana.herokuapp.com and setting localStorage.netlify = "deploy-preview-3448--cbioportalfrontend"

### CODE REVIEW

File: src/pages/studyView/addChartButton/AddChartButton.tsx [hunk 1]
```diff
@@ -303,6 +315,83 @@ class AddChartTabs extends React.Component<IAddChartTabsProps, {}> {
         }
     }
 
+    @computed get existSharedCustomData() {
+        if (this.props.store.isLoggedIn) {
```
Comment (alisman) [inline src/pages/studyView/addChartButton/AddChartButton.tsx:319]: is this condition correct?  no shared data if user IS logged in?

Comment (kalletlak) [reply src/pages/studyView/addChartButton/AddChartButton.tsx:319]: all shared data would be saved to user profile if he is logged in

File: src/pages/studyView/UserSelections.tsx [hunk 2]
```diff
@@ -270,6 +267,81 @@ export default class UserSelections extends React.Component<
...
+                ];
+                if (chartMeta) {
+                    chartMeta.clinicalAttribute!.datatype;
+                    if (
+                        chartMeta.clinicalAttribute!.datatype ===
+                        DataType.STRING
+                    ) {
+                        // Pie chart filter
+                        acc.push(
+                            <div className={styles.parentGroupLogic}>
```
Comment (Luke-Sikina) [inline src/pages/studyView/UserSelections.tsx:284]: Can you pull this out into its own method, or even its own component? The indentation is making this hard to read.

File: src/pages/studyView/UserSelections.tsx [hunk 3]
```diff
@@ -270,6 +267,81 @@ export default class UserSelections extends React.Component<
...
+                                            group={false}
+                                        />,
+                                    ]}
+                                    operation={':'}
+                                    group={false}
+                                />
+                            </div>
+                        );
+                    } else {
+                        // Bar chart filter
```
Comment (Luke-Sikina) [inline src/pages/studyView/UserSelections.tsx:336]: Add more to this comment to explain why the bar chart filter is `<></>`

File: src/pages/studyView/addChartButton/AddChartButton.tsx [hunk 4]
```diff
@@ -303,6 +315,83 @@ class AddChartTabs extends React.Component<IAddChartTabsProps, {}> {
...
+                        </span>{' '}
+                        are shared custom data.
+                    </>
+                );
+            }
+            // Notify that shared and page-session groups are not saved for non-logged users
+            if (
+                !this.props.store.isLoggedIn &&
+                AppConfig.serverConfig.authenticationMethod &&
+                AppConfig.serverConfig.authenticationMethod.includes(
```
Comment (Luke-Sikina) [inline src/pages/studyView/addChartButton/AddChartButton.tsx:350]: I think this makes 5 different places in the app where we make this check. Can we make a utility for this?

**Review by Luke-Sikina — APPROVED** [review]
Looks good to me. A few small suggestions, but nothing blocking.

File: src/pages/studyView/addChartButton/customCaseSelection/CustomCaseSelection.tsx [hunk 5]
```diff
@@ -87,17 +85,17 @@ export default class CustomCaseSelection extends React.Component<
...
                 : '',
             patientAttribute: this.caseIdsMode === ClinicalDataTypeEnum.PATIENT,
-            groups:
+            data:
                 this.result.validationResult.error.length === 0
-                    ? this.result.groups
+                    ? this.result.data
                     : [],
-        };
+        } as any;
```
Comment (alisman) [inline src/pages/studyView/addChartButton/customCaseSelection/CustomCaseSelection.tsx:98]: why as any?

File: src/pages/studyView/StudyViewPageStore.ts [hunk 6]
```diff
@@ -391,7 +378,26 @@ export class StudyViewPageStore {
             )
         );
 
-        // Include special charts into custom charts list
+        this.reactionDisposers.push(
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts:381]: this should have more comments to explain logic

File: src/pages/studyView/StudyViewPageStore.ts [hunk 7]
```diff
@@ -573,6 +617,31 @@ export class StudyViewPageStore {
         }
     }
 
+    //Save any shared custom data to user profile if user login after page is loaded
+    public async saveSharedCustomDataToUserProfile(
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts:621]: kinda minor but this method doesn't return anything so probably doesn't need to by async.

File: src/pages/studyView/StudyViewPageStore.ts [hunk 8]
```diff
@@ -4160,6 +4633,27 @@ export class StudyViewPageStore {
         }
     }
 
+    @autobind
+    @action
+    loadUserCustomData() {
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts:4638]: i don't the name of this method really describes what is going on.

File: src/pages/studyView/StudyViewPageStore.ts [hunk 9]
```diff
@@ -693,6 +762,128 @@ export class StudyViewPageStore {
...
+            isSharedChart?: boolean;
+        })[]
+    >({
+        invoke: async () => {
+            if (this.studyIds.length > 0 && this.isLoggedIn) {
+                const chustomChartsSessions = await sessionServiceClient.getCustomDataForStudies(
+                    this.studyIds.slice()
+                ); // slice because cant pass mobx
+
+                return chustomChartsSessions.map(chustomChartsSession => {
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts:777]: type "chustom"

File: src/pages/studyView/StudyViewPageStore.ts [hunk 10]
```diff
@@ -693,6 +762,128 @@ export class StudyViewPageStore {
...
+                        this.chartsType.set(uniqueKey, ChartTypeEnum.PIE_CHART);
+                        this.chartsDimension.set(uniqueKey, { w: 1, h: 1 });
+                    }
+                    this.customChartSet.set(uniqueKey, customChart);
+                });
+            }
+        },
+    });
+
+    readonly sharedCustomCharts = remoteData<
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts:821]: we need a comment here explaining what "shared" means.   i mean, obviously it means shared from another user, but also that (if i understand correctly) different things will happen with a shared group depending on logged in state of user.  we need to have all this explained at the very least in comments

**Review by alisman — APPROVED** [review]

**Review by dippindots — APPROVED** [review]
LGTM!

File: src/pages/studyView/StudyViewPage.tsx [hunk 11]
```diff
@@ -174,6 +176,15 @@ export default class StudyViewPage extends React.Component<
                         // Open group comparison manager if there are shared groups in the url
                         this.store.showComparisonGroupUI = true;
                     }
+                } else if (datum.startsWith('sharedCustomData')) {
```
Comment (zhx828) [inline src/pages/studyView/StudyViewPage.tsx:179]: Do you want to consider using a library to deal with hash then you can model them? https://github.com/sindresorhus/query-string#readme

**Review by zhx828 — COMMENTED** [review]
All minor comments, nothing blocks moving forward.

- After opening the shared link in a incognito mode, the shared chart does not show up https://cbioportal-fix-7726-0oimabktio.herokuapp.com/study/summary?id=acc_2019#sharedCustomData=5f986eade4b05f0d86ed4d3c

- Can we disable the Add button while session service is saving, so it would prevent accidental adding?

File: src/appShell/App/PortalHeader.tsx [hunk 12]
```diff
@@ -155,11 +155,8 @@ export default class PortalHeader extends React.Component<
                             <Else>
                                 <If
                                     condition={
-                                        AppConfig.serverConfig
-                                            .authenticationMethod &&
-                                        AppConfig.serverConfig.authenticationMethod.includes(
-                                            'social_auth'
-                                        )
```
[the remark this answers is not in the document]
Comment (kalletlak) [reply src/appShell/App/PortalHeader.tsx:162]: this is just a cleanup. I introduced `isSocialAuthenticated` in this pr and replaced it wherever `social_auth` is checked

**Comment by kalletlak** [comment]
> All minor comments, nothing blocks moving forward. Nice job 👍🏼
> 
> * After opening the shared link in a incognito mode, the shared chart does not show up https://cbioportal-fix-7726-0oimabktio.herokuapp.com/study/summary?id=acc_2019#sharedCustomData=5f986eade4b05f0d86ed4d3c

Seems working to me. Have you set `localStorage.netlify`?

> * Can we disable the Add button while session service is saving, so it would prevent accidental adding?

Yes, updated.

File: src/pages/studyView/StudyViewPage.tsx [hunk 13]
```diff
@@ -160,22 +163,23 @@ export default class StudyViewPage extends React.Component<
...
-                } else if (datum.startsWith('sharedGroups')) {
-                    const sharedGroups = datum.match(/sharedGroups=([^&]*)/);
-                    if (sharedGroups && sharedGroups.length > 1) {
-                        newStudyViewFilter.sharedGroups = sharedGroups[1];
-                        // Open group comparison manager if there are shared groups in the url
-                        this.store.showComparisonGroupUI = true;
-                    }
-                }
-            });
+            const params = parse(hashString);
```
Comment (zhx828) [inline src/pages/studyView/StudyViewPage.tsx:166]: Do you think it makes sense to type the hash object?

Comment (adamabeshouse) [reply src/pages/studyView/StudyViewPage.tsx:166]: +1 - `parse(hashString) as something`

Comment (kalletlak) [reply src/pages/studyView/StudyViewPage.tsx]: updated

**Review by adamabeshouse — APPROVED** [review]

**Comment by dippindots** [comment]
Hi @kalletlak, I noticed that this pr may has many conflicts with the pr I we just merged into master: https://github.com/cBioPortal/cbioportal-frontend/pull/3521. Please let me know if you need my help with resolving the conflict. Sorry about that.

**Comment by kalletlak** [comment]
> Hi @kalletlak, I noticed that this pr may has many conflicts with the pr I we just merged into master: #3521. Please let me know if you need my help with resolving the conflict. Sorry about that.

Should be ok. Thanks for letting me know

**Comment by alisman** [comment]
@kalletlak just reviewed this again and i think it's looking good from UI standpoint

**Pull request merged**